src/reconcile/rec_visibility.c

```diff
@@ -586,40 +587,64 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
+         * found" for reads after this point.
+         *
+         * Otherwise, leave the end of the visibility window at the maximum possible value to
+         * indicate that the value is visible to any timestamp/transaction id ahead of it.
          */
-        if ((upd_select->stop_ts == WT_TS_MAX && upd_select->stop_txn == WT_TXN_MAX) &&
-          ((upd_select->start_ts == WT_TS_NONE && upd_select->start_txn == WT_TXN_NONE) ||
-              __wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts))) {
-            upd_select->start_ts = WT_TS_NONE;
-            upd_select->start_txn = WT_TXN_NONE;
-            upd_select->stop_ts = WT_TS_MAX;
-            upd_select->stop_txn = WT_TXN_MAX;
+        WT_ASSERT(session, upd_select->upd != NULL);
+        if (upd_select->upd->type == WT_UPDATE_TOMBSTONE) {
+            upd_select->stop_ts = upd_select->upd->start_ts;
+            upd_select->stop_txn = upd_select->upd->txnid;
+            upd_select->upd = upd_select->upd->next;
+        }
+        if (upd_select->upd != NULL) {
+            /* The beginning of the validity window is the selected update's time pair. */
+            upd_select->durable_ts = upd_select->start_ts = upd_select->upd->start_ts;
+            upd_select->start_txn = upd_select->upd->txnid;
+        } else {
+            /*
+             * The only time we don't get the unpacked value is when we insert a key in which case
+             * we can't have a tombstone.
+             */
+            WT_ASSERT(session, vpack != NULL);
+            /*
+             * It's possible to have a tombstone as the only update in the update list. If we
+             * reconciled before with only a single update and then read the page back into cache,
+             * we'll have an empty update list. And applying a delete on top of that will result in
+             * ONLY a tombstone in the update list.
+             *
+             * In this case, we should leave the selected update unset to indicate that we want to
+             * keep the same on-disk value but set the stop time pair to indicate that the validity
+             * window ends when this tombstone started.
+             */
+            upd_select->durable_ts = upd_select->start_ts = vpack->start_ts;
+            upd_select->start_txn = vpack->start_txn;
+            /*
+             * Leaving the update unset means that we can skip reconciling. If we've set the stop
+             * time pair because of a tombstone after the on-disk value, we still have work to do so
+             * that is NOT ok. Let's allocate an update equivalent to the on-disk value and continue
+             * on our way!
+             */
+            WT_RET(__wt_scr_alloc(session, 0, &tmp));
+            WT_ERR(__wt_page_cell_data_ref(session, page, vpack, tmp));
+            WT_ERR(__wt_update_alloc(session, tmp, &upd_select->upd, &size, WT_UPDATE_STANDARD));
+            upd_select->upd->ext = 1;
```

Comment:
We're allocating 2 temporary buffers now (`key` and `tmp`), and I don't think the usage overlaps, perhaps we don't need both?

I'm also pretty sure `las_key` and `las_value` are misused -- there's an assumption that calling the LAS cursor for a key/value pair won't allocate space in the WT_ITEM, and I don't see any reason to believe that's guaranteed.

Reply:
Ok, we can use the same buffer.

What do you mean about misusing `las_key` and `las_value`? Are you saying that there is a missing free in the existing code?

Reply:
> Are you saying that there is a missing free in the existing code?

Yes. We clear a structure, then do `las_cursor->get_key/get_value` and there's no guarantee that `WT_CURSOR.get` not allocate memory in the `WT_ITEM`.

On second thought, I think that's true for `las_key` -- I think `las_value` is worse, there are calls to `__wt_buf_set` which, unless I'm missing something, guarantees we'll leak memory.

Reply:
Hi @keithbostic. I've had more of a think about this and I'd rather keep it as is.

My reasoning is that @quchenhao is currently working on a change to write the newest versions to disk. If this is the case, we're not going to have lookaside searching in this function which will remove the need for a `key` buffer. Also, the `las_key` and `las_value` will no longer exist.

So none of these issues will exist in the future but fixing them temporarily will make my conflicts harder to resolve because it will further tangle my patch with this version of the function.

Let me know if you don't agree with this or want to talk about it further.